flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/ProducerMergePartitionFileReader.java

```diff
@@ -21,38 +21,188 @@
+                subpartitionReadCache.computeIfAbsent(subpartitionId, ignore -> new HashMap<>());
+        Queue<SubpartitionReadCache> subpartitionProgresses =
+                progresses.computeIfAbsent(bufferIndex, ignore -> new LinkedList<>());
+        SubpartitionReadCache currentProgress =
+                subpartitionProgresses.isEmpty() ? null : subpartitionProgresses.peek();
+        if (currentProgress == null) {
+            checkState(bufferIndex == 0);
+            currentProgress = new SubpartitionReadCache(subpartitionId);
+            subpartitionProgresses.add(currentProgress);
+            progresses.put(0, subpartitionProgresses);
+        }
+
+        if (!currentProgress.hasBuffer()) {
+            return null;
+        }
+        long fileOffSet = currentProgress.getCurrentFileOffset();
+        if (fileChannel == null) {
+            try {
+                fileChannel = FileChannel.open(dataFilePath, StandardOpenOption.READ);
+            } catch (FileNotFoundException e) {
+                throw new PartitionNotFoundException(
+                        TieredStorageIdMappingUtils.convertId(partitionId));
+            }
+        }
+        try {
+            fileChannel.position(fileOffSet);
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(e, "Failed to position file offset to buffer.");
+        }
+        Buffer buffer = null;
+        try {
+            buffer = readFromByteChannel(fileChannel, reusedHeaderBuffer, segment, recycler);
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(e, "Failed to read buffer.");
+        }
+        currentProgress.advance(
+                checkNotNull(buffer).readableBytes() + BufferReaderWriterUtil.HEADER_LENGTH);
+        subpartitionProgresses.poll();
+        if (subpartitionProgresses.isEmpty()) {
+            progresses.remove(bufferIndex);
+        }
+        progresses
+                .computeIfAbsent(
+                        currentProgress.getCurrentBufferIndex(), ignore -> new LinkedList<>())
+                .add(currentProgress);
+        return buffer;
+    }
+
+    @Override
+    public long getPriority(
+            TieredStoragePartitionId partitionId,
+            TieredStorageSubpartitionId subpartitionId,
+            int segmentId,
+            int bufferIndex) {
+        Queue<SubpartitionReadCache> progress =
+                subpartitionReadCache
+                        .computeIfAbsent(subpartitionId, ignore -> new HashMap<>())
+                        .computeIfAbsent(bufferIndex, ignore -> new LinkedList<>());
+        return progress.isEmpty() ? 0 : progress.peek().getCurrentBufferIndex();
     }
 
     @Override
-    public void release() {}
+    public void release() {
+        fileChannel = null;
+        IOUtils.deleteFileQuietly(dataFilePath);
+    }
+
+    /**
+     * {@link SubpartitionReadCache} is the cache to record the reading progress for a consumer of a
+     * subpartition.
+     */
```

Comment:
Why is the partition file reader recording the reading progress of each consumer?